modules/lib/types-dag.nix

```diff
@@ -0,0 +1,96 @@
+{ dag, lib }:
+
+with lib;
+
+let
+
+  isDagEntry = e: isAttrs e && (e ? data) && (e ? after) && (e ? before);
+
+  dagContentType = elemType: types.submodule {
+    options = {
+      data = mkOption { type = elemType; };
+      after = mkOption { type = with types; uniq (listOf str); };
+      before = mkOption { type = with types; uniq (listOf str); };
+    };
+  };
+
+in
+
+{
+  # A directed acyclic graph of some inner type.
+  dagOf = elemType:
+    let
+      convertAllToDags =
+        let
+          maybeConvert = n: v:
+            if isDagEntry v
+            then v
+            else dag.entryAnywhere v;
+        in
+          map (def: def // { value = mapAttrs maybeConvert def.value; });
+
+      attrEquivalent = types.attrsOf (dagContentType elemType);
+    in
+      mkOptionType rec {
+        name = "dagOf";
+        description = "DAG of ${elemType.description}s";
+        check = x: dag.isDag x || isAttrs x;
```

Comment:
Should be equivalent to just `check = isAttrs` since `isDag` is a superset of it.

Or alternatively this could be a full check with

```nix
value: isAttrs value && all (x: isDagEntry x || elemType.check x) (attrValues value)
```

Reply:
Yeah, I added the `dag.isDag` in some attempt at being pedagogical 🙂 I've switched to just `isAttrs` for now. I might switch to the more in-depth check in the future.